src/Nethermind/Nethermind.Evm/Precompiles/Stateful/BeaconBlockRootPrecompile.cs

```diff
@@ -0,0 +1,77 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Specs;
+using Nethermind.Int256;
+using Nethermind.State;
+
+namespace Nethermind.Evm.Precompiles.Stateful;
+public class BeaconBlockRootPrecompile : IPrecompile
+{
+    private readonly IWorldState _state;
+    public static Address Address { get; } = Address.FromNumber(0x0B);
+    public static UInt256 HISTORICAL_ROOTS_LENGTH = 98304;
+    private BeaconBlockRootPrecompile(IWorldState state)
+    {
+        _state = state;
+    }
+
+    public static IPrecompile Instance(IWorldState state) => new BeaconBlockRootPrecompile(state);
+
+    private byte[] SloadFromStorage(in UInt256 index)
+    {
+        StorageCell storageCell = new(Address, index);
+        return _state.Get(storageCell);
+    }
+    public static void SetupBeaconStateRootPrecompileState(IWorldState stateProvider, Block block, Keccak ParentBeaconBlockRoot)
+    {
+        UInt256 timestamp = block.Header.Timestamp;
+
+        UInt256.Mod(timestamp, HISTORICAL_ROOTS_LENGTH, out UInt256 timestampReduced);
+        UInt256 rootIndex = timestampReduced + HISTORICAL_ROOTS_LENGTH;
+
+        StorageCell tsStorageCell = new(Address, timestampReduced);
+        StorageCell brStorageCell = new(Address, rootIndex);
+
+        stateProvider.Set(tsStorageCell, timestamp.ToBigEndian());
+        stateProvider.Set(brStorageCell, ParentBeaconBlockRoot.Bytes.ToArray());
+    }
+
+    public (ReadOnlyMemory<byte>, bool) Run(in ReadOnlyMemory<byte> inputData, IReleaseSpec releaseSpec)
+    {
+        Metrics.BeaconStatePrecompile++;
+
+        if (inputData.Length != 32)
+        {
+            return (Array.Empty<byte>(), false);
+        }
+
+        UInt256 timestamp = new UInt256(inputData.Span, true);
+        UInt256.Mod(timestamp, HISTORICAL_ROOTS_LENGTH, out UInt256 timestampReduced);
+        UInt256 recordedTimestamp = new UInt256(SloadFromStorage(timestampReduced), true);
+
+        if (recordedTimestamp != timestamp)
+        {
+            return (default, true);
+        }
+        else
+        {
+            UInt256 timestampExtended = timestampReduced + HISTORICAL_ROOTS_LENGTH;
+            byte[] recordedRoot = SloadFromStorage(timestampExtended);
+            return (recordedRoot, true);
+        }
+    }
+
+    public long BaseGasCost(IReleaseSpec releaseSpec)
+    {
+        return 4200;
```

Comment:
Do you think we should use 2*GastCostOf.Sload? Let's imagine that we will change the cost of SLOAD in the future.
`The precompile costs G_beacon_root gas to reflect the two (2) implicit SLOADs from the precompile’s state.`